go/vt/vtgate/semantics/bitset/bitset.go

```diff
@@ -0,0 +1,241 @@
+	for i := 0; i < min; i++ {
+		if bs[i]&b2[i] != 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// Or returns the logical OR of the two Bitsets as a new Bitset
+func (bs Bitset) Or(b2 Bitset) Bitset {
+	if len(bs) == 0 {
+		return b2
+	}
+	if len(b2) == 0 {
+		return bs
+	}
+
+	small, large := bs, b2
+	if len(small) > len(large) {
+		small, large = large, small
+	}
+
+	merged := make([]byte, len(large))
+	m := 0
+
+	for m < len(small) {
+		merged[m] = small[m] | large[m]
+		m++
+	}
+	for m < len(large) {
+		merged[m] = large[m]
+		m++
+	}
+	return toBitset(merged)
+}
+
+// AndNot returns the logical AND NOT of the two Bitsets as a new Bitset
+func (bs Bitset) AndNot(b2 Bitset) Bitset {
+	if len(b2) == 0 {
+		return bs
+	}
+
+	merged := make([]byte, len(bs))
+	m := 0
+
+	for m = 0; m < len(bs); m++ {
+		if m < len(b2) {
+			merged[m] = bs[m] & ^b2[m]
+		} else {
+			merged[m] = bs[m]
+		}
+	}
+	for ; m > 0; m-- {
+		if merged[m-1] != 0 {
+			break
+		}
+	}
+	return toBitset(merged[:m])
+}
+
+// And returns the logical AND of the two bitsets as a new Bitset
+func (bs Bitset) And(b2 Bitset) Bitset {
+	if len(bs) == 0 || len(b2) == 0 {
+		return ""
+	}
+
+	merged := make([]byte, minlen(bs, b2))
+	m := 0
+
+	for m = 0; m < len(merged); m++ {
+		merged[m] = bs[m] & b2[m]
+	}
+	for ; m > 0; m-- {
+		if merged[m-1] != 0 {
+			break
+		}
+	}
+	return toBitset(merged[:m])
+}
+
+// Set returns a copy of this Bitset where the bit at `offset` is set
+func (bs Bitset) Set(offset int) Bitset {
+	alloc := len(bs)
+	if max := bitsetWordSize(offset); max > alloc {
+		alloc = max
+	}
+
+	words := make([]byte, alloc)
+	copy(words, bs)
+	words[offset/bitsetWidth] |= 1 << (offset % bitsetWidth)
+	return toBitset(words)
+}
+
+// SingleBit returns the position of the single bit that is set in this Bitset
+// If the Bitset is empty, or contains more than one set bit, it returns -1
+func (bs Bitset) SingleBit() int {
+	offset := -1
+	for i := 0; i < len(bs); i++ {
+		t := bs[i]
+		if t == 0 {
+			continue
+		}
+		if offset >= 0 || bits.OnesCount8(t) != 1 {
+			return -1
+		}
+		offset = i*bitsetWidth + bits.TrailingZeros8(t)
+	}
+	return offset
+}
+
+// IsContainedBy returns whether this Bitset is contained by the given Bitset
+func (bs Bitset) IsContainedBy(b2 Bitset) bool {
+	if len(bs) > len(b2) {
+		return false
+	}
+	for i := 0; i < len(bs); i++ {
+		left := bs[i]
+		rigt := b2[i]
+		if left&rigt != left {
+			return false
+		}
+	}
+	return true
+}
+
+// Popcount returns the number of bits that are set in this Bitset
+func (bs Bitset) Popcount() (count int) {
+	for i := 0; i < len(bs); i++ {
+		count += bits.OnesCount8(bs[i])
+	}
+	return
+}
+
+// ForEach calls the given callback with the position of each bit set in this Bitset
+func (bs Bitset) ForEach(yield func(int)) {
+	for i := 0; i < len(bs); i++ {
+		bitset := bs[i]
+		for bitset != 0 {
+			t := bitset & -bitset
+			r := bits.TrailingZeros8(bitset)
+			yield(i*bitsetWidth + r)
+			bitset ^= t
+		}
+	}
+}
+
```

Comment:
an example of how this works will help. A general example for overall implementation will also work.